File: packages/interocitor/src/core/crdt.ts
/**
 * CRDT Merge Engine — configurable per-column merge strategy.
 *
 * Each column in each row carries its own HLC.
 * The merge strategy determines which value wins on conflict:
 *
 *  - `'remote-wins'` — Remote always overwrites local. (Default)
 *  - `'lww'`         — Last-Writer-Wins. Highest HLC wins.
 *  - `'local-wins'`  — Keep local value when both exist.
 *  - custom function  — `(local, remote, ctx) => ColumnEntry`
 *
 * Deletes are soft (tombstone with HLC) and always use LWW.
 */

import type {
  Row,
  Op,
  ColumnEntry,
  ChangeEntry,
  DatabaseSchemaDefinition,
  MergeStrategy,
  TableMergeConfig,
} from './types.ts';
import { hlcCompareStr } from './hlc.ts';

/** Reserved keys that are not user columns */
const META_KEYS = new Set(['_table', '_rowId', '_deleted', '_deletedHlc', '_schemaVersion']);

/**
 * Resolve the merge strategy for a specific column.
 *
 * Resolution order (first defined wins):
 *   field-level → table-level → database-level → 'lww'
 */
function resolveStrategy(
  schema: DatabaseSchemaDefinition | undefined,
  table: string,
  field: string,
): MergeStrategy {
  const tableDef = schema?.tables[table];
  if (tableDef?.merge) {
    const m = tableDef.merge;
    if (typeof m === 'object' && 'fields' in m) {
      // TableMergeConfig
      const config = m as TableMergeConfig;
      if (config.fields?.[field]) return config.fields[field];
      if (config.strategy) return config.strategy;
    } else {
      // bare MergeStrategy (string or function) on the table
      return m as MergeStrategy;
    }
  }
  // No schema at all → LWW (backwards compat for raw applyOp callers).
  // Schema present but no mergeStrategy → remote-wins (sensible default).
  if (!schema) return 'lww';
  return schema.mergeStrategy ?? 'remote-wins';
}

/**
 * Decide which column entry wins given a strategy.
 *
 * `local` may be undefined (new column). In that case, remote always wins
 * regardless of strategy — there's no conflict.
 */
function mergeColumn(
  local: ColumnEntry | undefined,
  remote: ColumnEntry,
  strategy: MergeStrategy,
  table: string,
  rowId: string,
  field: string,
): ColumnEntry | null {
  // No local value → accept remote unconditionally
  if (!local || !local.hlc) return remote;

  if (typeof strategy === 'function') {
    const result = strategy(local, remote, { table, rowId, field });
    // Only count as changed if the result differs from local
    return result.hlc !== local.hlc || result.value !== local.value ? result : null;
  }

  switch (strategy) {
    case 'remote-wins':
      return remote;

    case 'local-wins':
      // Only accept remote if it's strictly newer (no conflict — local
      // hasn't written this column yet at this HLC).
      // When both have values, local keeps its value.
      return null;

    case 'lww':
    default:
      return hlcCompareStr(remote.hlc, local.hlc) > 0 ? remote : null;
  }
}

/**
 * Apply a single op to the in-memory state.
 * Returns the affected row (mutated in place) or null if no change.
 */
export function applyOp(
  tables: Record<string, Record<string, Row>>,
  op: Op,
  schemaVersion: number,
  schema?: DatabaseSchemaDefinition,
): Row | null {
  if (!tables[op.table]) {
    tables[op.table] = {};
  }
  const table = tables[op.table];

  if (op.type === 'delete') {
    const existing = table[op.rowId];
    if (existing) {
      // Only apply delete if its HLC is newer than all column HLCs
      if (existing._deletedHlc && hlcCompareStr(op.hlc, existing._deletedHlc) <= 0) {
        return null; // stale delete
      }
      // Check if any column has a newer HLC than this delete
      const hasNewerColumn = Object.entries(existing).some(([key, val]) => {
        if (META_KEYS.has(key)) return false;
        const entry = val as ColumnEntry;
        return entry?.hlc && hlcCompareStr(entry.hlc, op.hlc) > 0;
      });
      if (hasNewerColumn) return null;

      existing._deleted = true;
      existing._deletedHlc = op.hlc;
      return existing;
    } else {
      // Tombstone for a row we haven't seen — create it
      const row: Row = {
        _table: op.table,
        _rowId: op.rowId,
        _deleted: true,
        _deletedHlc: op.hlc,
        _schemaVersion: schemaVersion,
      };
      table[op.rowId] = row;
      return row;
    }
  }

  // Upsert
  let row = table[op.rowId];
  let changed = false;

  if (!row) {
    row = {
      _table: op.table,
      _rowId: op.rowId,
      _deleted: false,
      _schemaVersion: schemaVersion,
    };
    table[op.rowId] = row;
    changed = true;
  }

  for (const [col, entry] of Object.entries(op.columns)) {
    const existing = row[col] as ColumnEntry | undefined;
    const strategy = resolveStrategy(schema, op.table, col);
    const winner = mergeColumn(existing, entry, strategy, op.table, op.rowId, col);
    if (winner) {
      row[col] = winner;
      changed = true;
    }
  }

  // An upsert that's newer than a delete revives the row
  if (row._deleted && row._deletedHlc) {
    const newestOpHlc = Object.values(op.columns).reduce((max, entry) => {
      return !max || hlcCompareStr(entry.hlc, max) > 0 ? entry.hlc : max;
    }, '' as string);

    if (newestOpHlc && hlcCompareStr(newestOpHlc, row._deletedHlc) > 0) {
      row._deleted = false;
      changed = true;
    }
  }

  return changed ? row : null;
}

/**
 * Apply a full change entry (potentially multiple ops).
 * Returns list of affected rows.
 */
export function applyChangeEntry(
  tables: Record<string, Record<string, Row>>,
  entry: ChangeEntry,
  schemaVersion: number,
  schema?: DatabaseSchemaDefinition,
): Row[] {
  const affected: Row[] = [];
  for (const op of entry.ops) {
    const row = applyOp(tables, op, schemaVersion, schema);
    if (row) affected.push(row);
  }
  return affected;
}

/**
 * Read a column value from a row, unwrapping the ColumnEntry.
 */
export function readColumn(row: Row, column: string): unknown {
  const entry = row[column];
  if (entry && typeof entry === 'object' && 'value' in entry && 'hlc' in entry) {
    return (entry as ColumnEntry).value;
  }
  return undefined;
}

/**
 * Build a plain object from a row (strip HLC metadata).
 */
export function rowToPlain(row: Row): Record<string, unknown> {
  const result: Record<string, unknown> = {
    _table: row._table,
    _rowId: row._rowId,
    _deleted: row._deleted,
  };
  for (const [key, val] of Object.entries(row)) {
    if (META_KEYS.has(key)) continue;
    if (val && typeof val === 'object' && 'value' in val && 'hlc' in val) {
      result[key] = (val as ColumnEntry).value;
    }
  }
  return result;
}

File: packages/interocitor-python/src/interocitor/crdt.py
"""Per-column CRDT merge rules used by Interocitor change entries."""

from __future__ import annotations

from collections.abc import Mapping
from typing import Any

from .hlc import hlc_compare_str
from .schema import MergeStrategy
from .types import ChangeEntry, ColumnEntry, DeleteOp, Op, Row, RowMeta, UpsertOp


Schema = Mapping[str, Any] | None


def _require_convergent_strategy(strategy: object, table: str, field: str) -> MergeStrategy:
    if strategy == "lww" or callable(strategy):
        return strategy  # type: ignore[return-value]
    raise ValueError(
        f"Unsupported replicated merge strategy {strategy!r} for {table}.{field}; "
        'use "lww" or a convergent custom merge'
    )


def _resolve_strategy(schema: Schema, table: str, field: str) -> MergeStrategy:
    """Implement core's field → table → database merge-strategy lookup."""

    if schema is None:
        # The TypeScript core intentionally uses LWW when there is no schema.
        return "lww"
    tables = schema.get("tables")
    table_def = tables.get(table) if isinstance(tables, Mapping) else None
    merge = table_def.get("merge") if isinstance(table_def, Mapping) else None
    if isinstance(merge, Mapping):
        # Core only recognizes a mapping as a TableMergeConfig when it carries
        # one of these keys.
        if "fields" in merge or "strategy" in merge:
            fields = merge.get("fields")
            field_strategy = fields.get(field) if isinstance(fields, Mapping) else None
            if field_strategy:
                return _require_convergent_strategy(field_strategy, table, field)
            strategy = merge.get("strategy")
            if strategy:
                return _require_convergent_strategy(strategy, table, field)
        else:
            return _require_convergent_strategy(merge, table, field)
    elif merge is not None:
        return _require_convergent_strategy(merge, table, field)
    strategy = schema.get("mergeStrategy")
    return "lww" if strategy is None else _require_convergent_strategy(strategy, table, field)


def _wire_value_equal(left: Any, right: Any) -> bool:
    """Compare decoded JSON values independently of object identity and key order."""

    if isinstance(left, bool) or isinstance(right, bool):
        return type(left) is type(right) and left is right
    left_is_number = isinstance(left, (int, float)) and not isinstance(left, bool)
    right_is_number = isinstance(right, (int, float)) and not isinstance(right, bool)
    if left_is_number or right_is_number:
        return left_is_number and right_is_number and left == right
    if isinstance(left, Mapping) or isinstance(right, Mapping):
        return (
            isinstance(left, Mapping)
            and isinstance(right, Mapping)
            and left.keys() == right.keys()
            and all(_wire_value_equal(left[key], right[key]) for key in left)
        )
    if isinstance(left, list) or isinstance(right, list):
        return (
            isinstance(left, list)
            and isinstance(right, list)
            and len(left) == len(right)
            and all(_wire_value_equal(a, b) for a, b in zip(left, right, strict=True))
        )
    return type(left) is type(right) and left == right


def _merge_column(
    local: ColumnEntry | None,
    remote: ColumnEntry,
    strategy: MergeStrategy,
    *,
    table: str,
    row_id: str,
    field: str,
) -> ColumnEntry | None:
    if local is None or not local.hlc:
        return remote
    if callable(strategy):
        result = strategy(local, remote, {"table": table, "rowId": row_id, "field": field})
        return result if result.hlc != local.hlc or not _wire_value_equal(result.value, local.value) else None
    comparison = hlc_compare_str(remote.hlc, local.hlc)
    if comparison > 0:
        return remote
    if comparison < 0:
        return None
    if not _wire_value_equal(local.value, remote.value):
        raise ValueError(f"Conflicting values share HLC {remote.hlc} at {table}/{row_id}.{field}")
    return None


def _blank_row(table: str, row_id: str, schema_version: int, *, deleted: bool = False, deleted_hlc: str | None = None) -> Row:
    return Row(
        _meta=RowMeta(
            table=table,
            row_id=row_id,
            deleted=deleted,
            deleted_hlc=deleted_hlc,
            schema_version=schema_version,
        ),
        payload={},
    )


def apply_op(
    tables: dict[str, dict[str, Row]],
    op: Op,
    schema_version: int,
    schema: Schema = None,
) -> Row | None:
    """Apply one operation in place and return the changed row, if any."""

    table = tables.setdefault(op.table, {})
    if isinstance(op, DeleteOp):
        existing = table.get(op.row_id)
        if existing is None:
            row = _blank_row(op.table, op.row_id, schema_version, deleted=True, deleted_hlc=op.hlc)
            table[op.row_id] = row
            return row
        if existing._meta.deleted_hlc and hlc_compare_str(op.hlc, existing._meta.deleted_hlc) <= 0:
            return None
        if any(entry.hlc and hlc_compare_str(entry.hlc, op.hlc) > 0 for entry in existing.payload.values()):
            return None
        existing._meta.deleted = True
        existing._meta.deleted_hlc = op.hlc
        existing.payload = {}
        return existing

    if not isinstance(op, UpsertOp):  # defensive guard if a third-party Op is supplied
        raise TypeError("Unknown Interocitor operation")
    row = table.get(op.row_id)
    changed = False
    if row is None:
        row = _blank_row(op.table, op.row_id, schema_version)
        table[op.row_id] = row
        changed = True

    columns = list(op.columns.items())
    if row._meta.deleted and row._meta.deleted_hlc:
        deleted_hlc = row._meta.deleted_hlc
        columns = [(name, entry) for name, entry in columns if hlc_compare_str(entry.hlc, deleted_hlc) > 0]
        if not columns:
            return None
        row.payload = {}
        row._meta.deleted = False
        row._meta.deleted_hlc = None
        changed = True

    for name, entry in columns:
        winner = _merge_column(
            row.payload.get(name),
            entry,
            _resolve_strategy(schema, op.table, name),
            table=op.table,
            row_id=op.row_id,
            field=name,
        )
        if winner is not None:
            row.payload[name] = winner
            changed = True
    return row if changed else None


def apply_change_entry(
    tables: dict[str, dict[str, Row]],
    entry: ChangeEntry,
    schema_version: int,
    schema: Schema = None,
) -> list[Row]:
    affected: list[Row] = []
    for operation in entry.ops:
        row = apply_op(tables, operation, schema_version, schema)
        if row is not None:
            affected.append(row)
    return affected


def read_column(row: Row, column: str) -> Any:
    entry = row.payload.get(column)
    return entry.value if entry is not None else None


def row_to_plain(row: Row) -> dict[str, Any]:
    value: dict[str, Any] = {
        "_table": row._meta.table,
        "_rowId": row._meta.row_id,
        "_deleted": row._meta.deleted,
    }
    value.update({name: entry.value for name, entry in row.payload.items()})
    return value
